README.md
Stop Motion Maker
=================

This application is supporting "Stop Motion(frame-by-frame) Animation" Making.

The blog is 
<a href="http://blog.orzlabs.org/search/label/StopMotionMaker">
http://blog.orzlabs.org/search/label/StopMotionMaker</a>

License
========
GPL v2.0. See LICENSE-GPL-2.0.txt



jni/CHANGELOG
18/03/03

1. (BUG*) reported size in -fsz mode was 8 bytes too small
2. (NEW) added -f switch for specifying the frame rate in fps
3. (BUG*) some DWORDs weren't being converted to little-endian
4. (NEW) added --version switch
5. *** VERSION 1.2 frozen ***


07/05/03

1. (BUG*) some C++-style comments in code
2. (BUG*) problem in byteswap.h for big-endian machines
3. (NEW) added FSF-recommended disclaimers
4. *** VERSION 1.3 frozen ***


13/10/03

1. (NEW) frame lists may be passed via STDIN; this corrects
   the previous problem--technically a kernel issue--of stack
   overflows due to really big frame lists
2. (CHANGE) made spacing a bit more K&Rish; really to give more
   consistency to the "wc -l" accounting of my projects
3. *** VERSION 1.4 frozen ***


07/01/04

1. (BUG*) frame lists on command line not working because
   linked-list doesn't get properly initialized
2. *** VERSION 1.5 frozen ***



jni/llist.c
/*
  llist.c

  routines for doubly-linked list

  Copyright (C) 2003 Phillip Bruce <dI77IHd@yahoo.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include "llist.h"
#include <stdlib.h>


/*
  spc: pulling l from list and deleting it from heap
  pre: l != 0
  pos: data elements still allocated; returning next
*/

List *list_erase(List *l)
{
  List *next = l->next;

  if (l->prev)
    l->prev->next = next;
  if (next)
    l->next->prev = l->prev;

  free(l);

  return next;
}


/*
  spc: pulling l from list, deleting it and its data
  pre: l != 0
  pos: returning l->next
*/

List *list_erased(List *l)
{
  if (l->data) {
    free(l->data);
    l->data = 0;
  }
  return list_erase(l);
}


/*
  spc: freeing all list structures
  pre: l != 0
  pos: data elements still allocated
*/

void list_rerase(List *l)
{
  if (l->next)
    list_rerase(l->next);
  l->next = 0;

  free(l);
}


/*
  spc: freeing all list structures and their data
  pre: l != 0
*/

void list_rerased(List *l)
{
  if (l->next)
    list_rerased(l->next);
  l->next = 0;

  if (l->data) {
    free(l->data);
    l->data = 0;
  }
  free(l);
}


/*
  spc: putting data in list at point l
*/

void list_insert(List *l, void *data)
{
  List *next = (List *) malloc(sizeof(List));
  next->data = l->data;
  next->next = l->next;
  next->prev = l;
  l->data = data;
  l->next = next;
}


/*
  spc: putting data on front of list
  pos: returning new front
*/

List *list_push_front(List *l, void *data)
{
  l = list_front(l);
  l->prev = (List *) malloc(sizeof(List));
  l->prev->data = data;
  l->prev->next = l;
  l->prev->prev = 0;

  return l->prev;
}


/*
  spc: putting data at end of list
  pos: returning new back
*/

List *list_push_back(List *l, void *data)
{
  l = list_back(l);
  l->next = (List *) malloc(sizeof(List));
  l->next->data = data;
  l->next->next = 0;
  l->next->prev = l;

  return l->next;
}


void list_pop_front(List *l)
{
  list_erase(list_front(l));
}


void list_pop_back(List *l)
{
  list_erase(list_back(l));
}


void list_popd_front(List *l)
{
  list_erased(list_front(l));
}


void list_popd_back(List *l)
{
  list_erased(list_back(l));
}


List *list_front(List *l)
{
  while (l->prev)
    l = l->prev;
  return l;
}


List *list_back(List *l)
{
  while (l->next)
    l = l->next;
  return l;
}


/*
  spc: removing elements l having l.data == data
  pos: data is not deleted
*/

void list_remove(List *l, void *data)
{
  List *l2;
  while ((l2 = list_find(l, data)))
    l = list_erase(l2); 
}


/*
  spc: finding next element l having l.data == data
*/

List *list_find(List *l, void *data)
{
  while (l && (l->data != data))
    l = l->next;
  return l;
}


/*
  spc: indexing l as an array
  pos: returning nth element, or 0 if out of range
*/

List *list_nth(List *l, unsigned int n) 
{
  int i;
  for (i = 0; (i < n) && (l); ++i, l = l->next); 
  return l;
}


unsigned int list_size(List *l)
{
  int i;
  for (i = 0; (l); ++i, l = l->next); 
  return i;
}



jni/llist.h
/*
  llist.h

  routines for doubly-linked list

  Copyright (C) 2003 Phillip Bruce <dI77IHd@yahoo.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#ifndef _LLIST_H_
#define _LLIST_H_


typedef struct _List List;


struct _List
{
  void *data;
  List *next;
  List *prev;
};


/*
  n.b. I've loosely followed the C++/STL naming conventions

  by definition, the back is the List l s.t. l.next == 0
                 the front is the List l s.t. l.prev == 0


  all functions assuming l to be a valid pointer

  in general, there exist "d" version to destroyer functions
  that also free data
*/

List        *list_erase      (List *l);
List        *list_erased     (List *l);
void         list_rerase     (List *l);
void         list_rerased    (List *l);
void         list_insert     (List *l, void *data);
List        *list_push_front (List *l, void *data);
List        *list_push_back  (List *l, void *data);
void         list_pop_front  (List *l);
void         list_pop_back   (List *l);
void         list_popd_front (List *l);
void         list_popd_back  (List *l);
List        *list_front      (List *l);
List        *list_back       (List *l);

void         list_remove     (List *l, void *data);
List        *list_find       (List *l, void *data);
List        *list_nth        (List *l, unsigned int n);
unsigned int list_size       (List *l);



#endif


jni/avifmt.h
#ifndef _AVIFMT_H_
#define _AVIFMT_H_


/* 4 bytes */
typedef int WORD;
typedef unsigned int DWORD;
/* 1 byte */
typedef char BYTE;


/* for use in AVI_avih.flags */
const DWORD AVIF_HASINDEX = 0x00000010;	/* index at end of file */
const DWORD AVIF_MUSTUSEINDEX = 0x00000020;
const DWORD AVIF_ISINTERLEAVED = 0x00000100;
const DWORD AVIF_TRUSTCKTYPE = 0x00000800;
const DWORD AVIF_WASCAPTUREFILE = 0x00010000;
const DWORD AVIF_COPYRIGHTED = 0x00020000;


struct AVI_avih 
{
  DWORD	us_per_frame;	/* frame display rate (or 0L) */
  DWORD max_bytes_per_sec;	/* max. transfer rate */
  DWORD padding;	/* pad to multiples of this size; */
  /* normally 2K */
  DWORD flags;
  DWORD tot_frames;	/* # frames in file */
  DWORD init_frames;
  DWORD streams;
  DWORD buff_sz;  
  DWORD width;
  DWORD height;
  DWORD reserved[4];
};


struct AVI_strh
{
  unsigned char type[4];      /* stream type */
  unsigned char handler[4];
  DWORD flags;
  DWORD priority;
  DWORD init_frames;       /* initial frames (???) */
  DWORD scale;
  DWORD rate;
  DWORD start;
  DWORD length;
  DWORD buff_sz;           /* suggested buffer size */
  DWORD quality;
  DWORD sample_sz;
};


struct AVI_strf
{       
  DWORD sz;
  DWORD width;
  DWORD height;
  DWORD planes_bit_cnt;
  unsigned char compression[4];
  DWORD image_sz;
  DWORD xpels_meter;
  DWORD ypels_meter;
  DWORD num_colors;        /* used colors */
  DWORD imp_colors;        /* important colors */
  /* may be more for some codecs */
};


/*
  AVI_list_hdr

  spc: a very ubiquitous AVI struct, used in list structures
       to specify type and length
*/

struct AVI_list_hdr 
{
  unsigned char id[4];   /* "LIST" */
  DWORD sz;              /* size of owning struct minus 8 */
  unsigned char type[4]; /* type of list */
};


struct AVI_list_odml 
{
  struct AVI_list_hdr list_hdr;

  unsigned char id[4];
  DWORD sz;
  DWORD frames;
};


struct AVI_list_strl 
{
  struct AVI_list_hdr list_hdr;
  
  /* chunk strh */
  unsigned char strh_id[4];
  DWORD strh_sz;
  struct AVI_strh strh;

  /* chunk strf */
  unsigned char strf_id[4];
  DWORD strf_sz;
  struct AVI_strf strf;

  /* list odml */
  struct AVI_list_odml list_odml;
};


struct AVI_list_hdrl 
{
  struct AVI_list_hdr list_hdr;

  /* chunk avih */
  unsigned char avih_id[4];
  DWORD avih_sz;
  struct AVI_avih avih;
  
  /* list strl */
  struct AVI_list_strl strl;
};


#endif


src/org/orzlabs/java/media/AviWriter.java
package org.orzlabs.java.media;

import android.util.Log;


public class AviWriter {
	static {
		System.loadLibrary("jpegtoavi-jni");
	}
	private String[] args;
	private int argsLength;
	public AviWriter(String[] arguments) {
		// TODO Auto-generated constructor stub
		args = arguments;
		argsLength = arguments.length;
	}
	/**
	 * @param args
	 */
	public static void main(String[] args) {
		AviWriter j2at = new AviWriter(args);
		j2at.execute();
	}
	public void execute() {
		Log.d("AviWriter", "retcode:" + jpegToAviJNI()); 
//		// TODO Auto-generated method stub
//		File outFile = new File(args[args.length - 1]);
//		try {
//			AviWriter aviWriter = new AviWriter(outFile, 2, true);
//			
//			for (int i = 0; i < args.length - 1; i++) {
//				File inFile = new File(args[i]);
//				FileInputStream fis = new FileInputStream(inFile);
//				long size = inFile.length();
//				System.out.println("size:" + size);
//				ByteArrayOutputStream baos = new ByteArrayOutputStream();
//				byte[] buffer = new byte[2048];
//				byte[] trash = new byte[10];
//				
////				fis.read(buffer, 0, 6);
////				baos.write(buffer, 5, 1);
////				for (int j = 0; j < buffer.length; j++) {
////					System.out.print(buffer[j]);
////				}
////				//fis.read(trash, 0, 5);
////				fis.read(trash, 0, 4);
////				byte[] avi1 = {'A', 'V', 'I', '1'};
////				baos.write(avi1);
//				int readSize;
//				while ((readSize = fis.read(buffer, 0, 2048)) != -1) {
//					baos.write(buffer, 0, readSize);
//				}
//				aviWriter.writeFrame(baos.toByteArray());
//				fis.close();
//			}
//			aviWriter.setFramesPerSecond(1, 1);
//			aviWriter.setSamplesPerSecond(1);
//			aviWriter.setDimensions(1, 1);
//			aviWriter.close();
//		} catch (IOException e) {
//			// TODO Auto-generated catch block
//			e.printStackTrace();
//		}
	}
	
	public native int jpegToAviJNI();


}


gen/org/orzlabs/android/stopmotionmaker/R.java
/* AUTO-GENERATED FILE.  DO NOT MODIFY.
 *
 * This class was automatically generated by the
 * aapt tool from the resource data it found.  It
 * should not be modified by hand.
 */

package org.orzlabs.android.stopmotionmaker;

public final class R {
    public static final class attr {
    }
    public static final class drawable {
        public static final int icon=0x7f020000;
    }
    public static final class id {
        public static final int handpreview=0x7f050002;
        public static final int shutter=0x7f050000;
        public static final int surfaceview=0x7f050001;
    }
    public static final class layout {
        public static final int main=0x7f030000;
    }
    public static final class string {
        public static final int Start=0x7f040002;
        public static final int app_name=0x7f040001;
        public static final int hello=0x7f040000;
    }
}


src/org/orzlabs/android/stopmotionmaker/StopMotionMaker.java
package org.orzlabs.android.stopmotionmaker;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

import org.orzlabs.java.media.AviWriter;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.hardware.Camera;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

public class StopMotionMaker extends Activity {
	private static final int IN_SAMPLE_SIZE = 1;
	private static final String TAG = "StopMotionMaker";

	private Camera mCamera;
	private ImageView mImage;
	private boolean mInProgress;
	private MediaRecorder recorder;
	private SurfaceHolder surfaceHolder;

//	private SurfaceHolder.Callback mSurfaceListener =
//		new SurfaceHolder.Callback() {
//
//		@Override
//		public void surfaceCreated(SurfaceHolder holder) {
//			mCamera = Camera.open();
//			Log.i(TAG, "Camera opened");
//			try {
//				mCamera.setPreviewDisplay(holder);
//			}
//			catch (Exception e) {
//				e.printStackTrace();
//			}
//		}
//
//		@Override
//		public void surfaceDestroyed(SurfaceHolder holder) {
//			mCamera.release();
//			mCamera = null;
//			Log.i(TAG, "Camera released");
//		}
//
//		@Override
//		public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
//			Camera.Parameters parameters = mCamera.getParameters();
//			parameters.setPreviewSize(width, height);
//			mCamera.setParameters(parameters);
//			mCamera.startPreview();
//			Log.i(TAG, "Camera preview started");
//			Log.i(TAG, "width  : " + width);
//			Log.i(TAG, "height : " + height);
//		}
//	};
//
//	private Camera.AutoFocusCallback mAutoFocusListener =
//		new Camera.AutoFocusCallback() {  
//		@Override
//		public void onAutoFocus(boolean success, Camera camera) {
//			Log.i(TAG,"AutoFocus : " + success);
//			camera.autoFocus(null);
//			camera.takePicture(mShutterListener, null, mPictureListener);
//			mInProgress = true;
//		}
//	};
//
//	private Camera.ShutterCallback mShutterListener = 
//		new Camera.ShutterCallback() {
//		@Override
//		public void onShutter() {
//			Log.i(TAG, "onShutter");
//		}
//	};
//
//	private View.OnClickListener mButtonListener = 
//		new View.OnClickListener() {   
//
//		@Override
//		public void onClick(View v) {
//			try {
//				recorder = new MediaRecorder();
////				recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
////				recorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
////				recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
//				recorder.setVideoSource(MediaRecorder.VideoSource.CAMERA);
//				recorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
//				recorder.setVideoEncoder(MediaRecorder.VideoEncoder.MPEG_4_SP);
//				recorder.setVideoFrameRate(15);
//				recorder.setVideoSize(176, 144);
//				File saveDir = Environment.getExternalStorageDirectory();
//				File saveFile = File.createTempFile("test", ".mp4", saveDir);
//				recorder.setOutputFile(saveFile.getAbsolutePath());
//				recorder.setPreviewDisplay(surfaceHolder.getSurface());
//
//				recorder.prepare();
//				recorder.start();
//				sleep();
//				recorder.stop();
//				recorder.release();
//			} catch (IOException e) {
//				Log.e(TAG, e.toString(), e);
//			}
//			//			if(mCamera != null && mInProgress == false){
//			//				mCamera.autoFocus(mAutoFocusListener);
//			//			}
//		}
//
//		private void sleep() {
//			try {
//				Thread.sleep(1000L);
//			} catch (InterruptedException e) {
//				Log.w(TAG, e);
//			}
//		}
//	};
//
//	private Camera.PictureCallback mPictureListener = 
//		new Camera.PictureCallback() {   
//		@Override
//		public void onPictureTaken(byte[] data, Camera camera) {
//			Log.i(TAG, "Picture taken");
//			if(data != null) {
//				Log.i(TAG, "JPEG Picture Taken");
//				BitmapFactory.Options options = new BitmapFactory.Options();
//				options.inSampleSize = IN_SAMPLE_SIZE;
//				Bitmap bitmap = BitmapFactory.decodeByteArray(data, 0, data.length, options);
//
//				mImage.setImageBitmap(bitmap);
//				mImage.setVisibility(View.VISIBLE);
//				try {
//					data2sd(data, "test.jpg");
//				} catch (IOException e) {
//					Log.e(this.toString(), e.getMessage(), e);
//				}
//
//				camera.startPreview();
//				mInProgress = false;
//			}
//		}
//	};
	private void data2sd(byte[] data, String fileName)
	throws IOException {
		FileOutputStream fos = null;
		try {
			fos = new FileOutputStream("/sdcard/" + fileName);
			fos.write(data);
		} finally {
			if (fos != null) {
				fos.close();
			}
		}
	}
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);

		File outFile = new File("/sdcard/hoge.avi");
		AviWriter aviWriter;
		try {
			String[] args = {
					"jpegtoavi",
					"-f", "1", "480", "640",
					"sdcard/DCIM/100ANDRO/DSC_0001.jpg",
					"sdcard/DCIM/100ANDRO/DSC_0002.jpg",
					"sdcard/DCIM/100ANDRO/DSC_0003.jpg",
			};
			Log.d(TAG, args[0] + ", " + args[1]);
			aviWriter = new AviWriter(args);
			aviWriter.execute();
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
//		mImage = (ImageView) findViewById(R.id.handpreview);
//		SurfaceView surface = (SurfaceView) findViewById(R.id.surfaceview);
//		surfaceHolder = surface.getHolder();
//
//		surfaceHolder.addCallback(mSurfaceListener);
//		surfaceHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS);
//
//		Button button = (Button) findViewById(R.id.shutter);
//		button.setOnClickListener(mButtonListener);
	} 

}

jni/jpegtoavi.c
/*
  jpegtoavi

  A simple converter of JPEGs to an AVI/MJPEG animation.

  Copyright (C) 2003 Phillip Bruce <dI77IHd@yahoo.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#define _LARGEFILE_SOURCE
#define _LARGEFILE64_SOURCE


#include "byteswap.h"
#include <stdio.h>
#include "avifmt.h"
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include "llist.h"
#include <jni.h>
#include <android/log.h>


/*
  containing name of jpeg file, size in
  bytes, and offset in pending avi file;

  we create a list of these for later
  production of the jpeg data and indices
  in the avi
 */

typedef struct _Jpeg_Data Jpeg_Data;

struct _Jpeg_Data
{
	DWORD size;
	DWORD offset;
	char name[0]; /* i.e. variable length structure */
};


#define JPEG_DATA_SZ (sizeof(DWORD) * 2)


/*
  spc: indicating file sz in bytes, -1 on error
 */

off_t file_sz(char *fn)
{
	struct stat s;
	if (stat(fn, &s) == -1)
		return -1;
	return s.st_size;
}


/*
  spc: returning sum of sizes of named JPEGs, -1 on error;
       file sizes adjusted to multiple of 4-bytes
  pos: l->size fields set to true file size
 */

off_t get_file_sz(List *l)
{
	off_t tmp, ret = 0;

	for(; (l); l = l->next) {
#if VERBOSE >= 2
		printf("determining file size for file == %s\n",
				((Jpeg_Data *) l->data)->name);
#endif

		if((tmp = file_sz(((Jpeg_Data *)l->data)->name)) == -1)
			return -1;
		((Jpeg_Data *)l->data)->size = (DWORD) tmp;
		tmp += ((4-(tmp%4)) % 4);
		ret += tmp;
	}

	return ret;
}


/* 
   spc: for obtaining list of file names from STDIN 
 */

List *get_file_list_stdin(void)
{
	char fn[PATH_MAX];
	Jpeg_Data *tmp;
	List *ret = (List *) malloc(sizeof(List)),
			*l = ret;

	ret->data = 0;
	ret->prev = 0;
	ret->next = 0;

	while (scanf("%s", fn) == 1) {
#if VERBOSE >= 2
		printf("read file name %s on stdin\n", fn);
#endif

		tmp = (Jpeg_Data *) malloc(strlen(fn) + 1 + JPEG_DATA_SZ);
		tmp->offset = 0;
		tmp->size = 0;
		strcpy(tmp->name, fn);
		if (l->data == 0)
			l->data = tmp;
		else
			l = list_push_back(l, tmp);
	}

#if VERBOSE >= 2
	printf("list size == %d\n", list_size(ret));
#endif

	return ret;
}


/*
  spc: for obtaining list of file names from argv
 */

List *get_file_list_argv(int argc, char **argv)
{
	List *ret = (List *) malloc(sizeof(List)),
			*l = ret;
	Jpeg_Data *tmp;
	int i;

	ret->data = 0;
	ret->prev = 0;
	ret->next = 0;

	for (i = 0; i < argc; ++i) {
		tmp = (Jpeg_Data *) malloc(strlen(argv[i]) + 1 + JPEG_DATA_SZ);
		strcpy(tmp->name, argv[i]);
		if (l->data == 0)
			l->data = tmp;
		else
			l = list_push_back(l, tmp);
	}

	return ret;
}


/*
  spc: printing 4 byte word in little-endian fmt
 */

void print_quartet(unsigned int i, FILE *out_fp)
{
	putc(i % 0x100, out_fp);  i /= 0x100;
	putc(i % 0x100, out_fp);  i /= 0x100;
	putc(i % 0x100, out_fp);  i /= 0x100;
	putc(i % 0x100, out_fp);
}


void show_help(int argc, char **argv)
{
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "jpegtoavi v%d.%d\n\
			Copyright (C) 2003 Phillip Bruce\n\
			\n\
			This is free software; see the source for copying conditions.  There is NO\n\
			warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n\
			USAGE: jpegtoavi {usec per img} {img width} {img height} {img1 .. imgN}\n\
			jpegtoavi -f {fps} {img width} {img height} {img1 .. imgN}\n\
			jpegtoavi -fsz {img1 .. imgN}\n\n\
			Creating an AVI/MJPEG on stdout.\n", 1, 5);
}

int
Java_org_orzlabs_java_media_AviWriter_jpegToAviJNI( JNIEnv* env,
		jobject thiz)
//int main(int argc, char **argv)
{
	int called_no = 0;
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "called.", called_no++);
	DWORD per_usec = 1;
	DWORD width;
	DWORD height;
	DWORD frames = 1;
	unsigned int fps;
	unsigned short img0;
	off64_t jpg_sz_64, riff_sz_64;
	long jpg_sz = 1;
	const off64_t MAX_RIFF_SZ = 2147483648LL; /* 2 GB limit */
	DWORD riff_sz;
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "called.", called_no++);

	int fd;
	long nbr, nbw, tnbw = 0;
	char buff[512];
	off_t mfsz, remnant;
	List *frlst = 0,
			*f = 0;
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "called.", called_no++);

	jclass clazz = (*env)->GetObjectClass(env, thiz);
	jfieldID fid = (*env)->GetFieldID(env ,clazz, "args", "[Ljava/lang/String;");
	jobjectArray *objStrings = (*env)->GetObjectField(env, thiz, fid);
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "called.", called_no++);

	int argc = (*env)->GetArrayLength(env, objStrings);
	char *argv[argc];
	int i;
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d, argc:%d\n", "called.", called_no++, argc);
	for (i = 0; i < argc; i++) {
		jstring text = (*env)->GetObjectArrayElement(env, objStrings, i);
		argv[i] = (char*) (*env)->GetStringUTFChars(env, text, NULL);
		__android_log_print(ANDROID_LOG_DEBUG, "JNI", "argv[%d]:%s", i, argv[i]);
	}
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "called.", called_no++);

	struct AVI_list_hdrl hdrl = {
			/* header */
			{
					{'L', 'I', 'S', 'T'},
					LILEND4(sizeof(struct AVI_list_hdrl) - 8),
					{'h', 'd', 'r', 'l'}
			},

			/* chunk avih */
			{'a', 'v', 'i', 'h'},
			LILEND4(sizeof(struct AVI_avih)),
			{
					LILEND4(per_usec),
					LILEND4(1000000 * (jpg_sz/frames) / per_usec),
					LILEND4(0),
					LILEND4(AVIF_HASINDEX),
					LILEND4(frames),
					LILEND4(0),
					LILEND4(1),
					LILEND4(0),
					LILEND4(width),
					LILEND4(height),
					{LILEND4(0), LILEND4(0), LILEND4(0), LILEND4(0)}
			},

			/* list strl */
			{
					{
							{'L', 'I', 'S', 'T'},
							LILEND4(sizeof(struct AVI_list_strl) - 8),
							{'s', 't', 'r', 'l'}
					},

					/* chunk strh */
					{'s', 't', 'r', 'h'},
					LILEND4(sizeof(struct AVI_strh)),
					{
							{'v', 'i', 'd', 's'},
							{'M', 'J', 'P', 'G'},
							LILEND4(0),
							LILEND4(0),
							LILEND4(0),
							LILEND4(per_usec),
							LILEND4(1000000),
							LILEND4(0),
							LILEND4(frames),
							LILEND4(0),
							LILEND4(0),
							LILEND4(0)
					},

					/* chunk strf */
					{'s', 't', 'r', 'f'},
					sizeof(struct AVI_strf),
					{
							LILEND4(sizeof(struct AVI_strf)),
							LILEND4(width),
							LILEND4(height),
							LILEND4(1 + 24*256*256),
							{'M', 'J', 'P', 'G'},
							LILEND4(width * height * 3),
							LILEND4(0),
							LILEND4(0),
							LILEND4(0),
							LILEND4(0)
					},

					/* list odml */
					{
							{
									{'L', 'I', 'S', 'T'},
									LILEND4(16),
									{'o', 'd', 'm', 'l'}
							},
							{'d', 'm', 'l', 'h'},
							LILEND4(4),
							LILEND4(frames)
					}
			}
	};

	/* parsing command line arguments */
	if (argc < 2) {
		show_help(argc, argv);
		__android_log_print(ANDROID_LOG_DEBUG, "JNI", "argc < 2");

		return -1;

	} else if (argc == 2 && strcmp(argv[1], "--version") == 0) {
		printf("jpegtoavi v%d.%d\n", 1, 5);
		return 0;

	} else if (argc < 3) {
		show_help(argc, argv);
		__android_log_print(ANDROID_LOG_DEBUG, "JNI", "argc < 3");

		return -1;

	} else if (strcmp(argv[1], "-fsz") == 0) {
		img0 = 2;

	} else if (strcmp(argv[1], "-f") == 0) {
		if (argc < 5 || sscanf(argv[2], "%u", &fps) != 1
				|| fps == 0
				|| sscanf(argv[3], "%u", &width) != 1
				|| sscanf(argv[4], "%u", &height) != 1) {
			show_help(argc, argv);
			__android_log_print(ANDROID_LOG_DEBUG, "JNI", "argc < 5, other");

			return -1;

		} else {
			per_usec = 1000000 / fps;
			img0 = 5;
		}

	} else if(argc < 4 || sscanf(argv[1], "%u", &per_usec) != 1
			|| sscanf(argv[2], "%u", &width) != 1
			|| sscanf(argv[3], "%u", &height) != 1) {
		show_help(argc, argv);
		__android_log_print(ANDROID_LOG_DEBUG, "JNI", "argc < 4, other");

		return -1;

	} else {
		img0 = 4;
	}

	if (img0 >= argc)
		frlst = get_file_list_stdin();
	else
		frlst = get_file_list_argv(argc - img0 , &argv[img0]);

	frames = list_size(frlst);

	/* getting image, and hence, riff sizes */
	jpg_sz_64 = get_file_sz(frlst);

	if (jpg_sz_64 == -1) {
		fprintf(stderr, "couldn't determine size of images\n");
		return -2;
	}

	riff_sz_64 = sizeof(struct AVI_list_hdrl) + 4 + 4 + jpg_sz_64
			+ 8*frames + 8 + 8 + 16*frames;

	if (riff_sz_64 >= MAX_RIFF_SZ) {
		fprintf(stderr,"RIFF would exceed 2 Gb limit\n");
		return -3;
	}

	jpg_sz = (long) jpg_sz_64;
	riff_sz = (DWORD) riff_sz_64;

	/* printing predicted file size and quitting */
	if (img0 == 2) {
		__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%lu\n", (unsigned long) riff_sz + 8UL);
		return 0;
	}
	FILE *out_fp;
	char *out_file_path = "/sdcard/DCIM/100ANDRO/test.avi";
	out_fp = fopen(out_file_path, "w");
	if (out_fp == NULL) {
		__android_log_print(ANDROID_LOG_ERROR, "JNI", "file open error: %s\n", out_file_path);
		return -1;
	}

	/* printing AVI riff hdr */
	fprintf(out_fp, "RIFF");
	print_quartet(riff_sz, out_fp);
	fprintf(out_fp, "AVI ");

	/* list hdrl */
	hdrl.avih.us_per_frame = LILEND4(per_usec);
	hdrl.avih.max_bytes_per_sec = LILEND4(1000000 * (jpg_sz/frames)
			/ per_usec);
	hdrl.avih.tot_frames = LILEND4(frames);
	hdrl.avih.width = LILEND4(width);
	hdrl.avih.height = LILEND4(height);
	hdrl.strl.strh.scale = LILEND4(per_usec);
	hdrl.strl.strh.rate = LILEND4(1000000);
	hdrl.strl.strh.length = LILEND4(frames);
	hdrl.strl.strf.width = LILEND4(width);
	hdrl.strl.strf.height = LILEND4(height);
	hdrl.strl.strf.image_sz = LILEND4(width * height * 3);
	hdrl.strl.list_odml.frames = LILEND4(frames);	/*  */
	fwrite(&hdrl, sizeof(hdrl), 1, out_fp);

	/* list movi */
	fprintf(out_fp, "LIST");
	print_quartet(jpg_sz + 8*frames + 4, out_fp);
	fprintf(out_fp, "movi");

	/* list movi .. frames */
	for (f = frlst; (f); f = f->next) {
		fprintf(out_fp, "00db");
		mfsz = ((Jpeg_Data *) f->data)->size;
		remnant = (4-(mfsz%4)) % 4;
		print_quartet(mfsz + remnant, out_fp);
		((Jpeg_Data *) f->data)->size += remnant;

		if (f == frlst) {
			((Jpeg_Data *) f->data)->offset = 4;

		} else {
			((Jpeg_Data *) f->data)->offset =
					((Jpeg_Data *) f->prev->data)->offset +
					((Jpeg_Data *) f->prev->data)->size + 8;
		}

		if ((fd = open(((Jpeg_Data *) f->data)->name, O_RDONLY)) < 0) {
			fprintf(stderr, "couldn't open file!\n");
			list_rerased(frlst);
			return -6;
		}
		nbw = 0;

		if ((nbr = read(fd, buff, 6)) != 6) {
			fprintf(stderr, "error\n");
			list_rerased(frlst);
			return -7;
		}
		fwrite(buff, nbr, 1, out_fp);
		read(fd, buff, 4);
		fwrite("AVI1", 4, 1, out_fp);
		nbw = 10;

		while ((nbr = read(fd, buff, 512)) > 0) {
			fwrite(buff, nbr, 1, out_fp);
			nbw += nbr;
		}
		if (remnant > 0) {
			fwrite(buff, remnant, 1, out_fp);
			nbw += remnant;
		}
		tnbw += nbw;
		close(fd);
	}

	if (tnbw != jpg_sz) {
		fprintf(stderr, "error writing images (wrote %ld bytes, expected %ld bytes)\n",
				tnbw, jpg_sz);
		list_rerased(frlst);
		return -8;
	}

	/* indices */
	fprintf(out_fp, "idx1");
	print_quartet(16 * frames, out_fp);
	for (f = frlst; (f); f = f->next) {
		fprintf(out_fp, "00db");
		print_quartet(18, out_fp);
		print_quartet(((Jpeg_Data *) f->data)->offset, out_fp);
		print_quartet(((Jpeg_Data *) f->data)->size, out_fp);
	}

	close(fd);
	fclose(out_fp);
	list_rerased(frlst);
	__android_log_print(ANDROID_LOG_DEBUG, "JNI", "%s:%d\n", "success.", called_no++);

	return 0;
}
